store/storage/pebbledb/db.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	return &Database{
+		storage: db,
+	}, nil
+}
+
+func NewWithDB(storage *pebble.DB) *Database {
+	return &Database{
+		storage: storage,
+	}
+}
+
+func (db *Database) Close() error {
+	return db.storage.Close()
+}
+
+func (db *Database) SetLatestVersion(version uint64) error {
+	var ts [VersionSize]byte
+	binary.LittleEndian.PutUint64(ts[:], version)
+	return db.storage.Set([]byte(latestVersionKey), ts[:], defaultWriteOpts)
+}
+
+func (db *Database) GetLatestVersion() (uint64, error) {
+	bz, closer, err := db.storage.Get([]byte(latestVersionKey))
+	if err != nil {
+		return 0, err
+	}
+
+	if len(bz) == 0 {
+		return 0, closer.Close()
+	}
+
+	return binary.LittleEndian.Uint64(bz), closer.Close()
+}
+
+func (db *Database) Has(storeKey string, version uint64, key []byte) (bool, error) {
+	_, err := getMVCCSlice(db.storage, storeKey, key, version)
+	if err != nil {
+		if errors.Is(err, store.ErrRecordNotFound) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("failed to perform PebbleDB read: %w", err)
+	}
+
+	return true, nil
+}
+
+func (db *Database) Get(storeKey string, version uint64, key []byte) ([]byte, error) {
+	bz, err := getMVCCSlice(db.storage, storeKey, key, version)
+	if err != nil {
+		if errors.Is(err, store.ErrRecordNotFound) {
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("failed to perform PebbleDB read: %w", err)
+	}
+
+	return bz, nil
+}
+
+func (db *Database) Set(storeKey string, version uint64, key, value []byte) error {
+	var versionBz [VersionSize]byte
+	binary.LittleEndian.PutUint64(versionBz[:], version)
+
+	batch := db.storage.NewBatch()
+
+	if err := batch.Set([]byte(latestVersionKey), versionBz[:], nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+	if err := batch.Set(MVCCEncode(prependStoreKey(storeKey, key), version), value, nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+
+	if err := batch.Commit(defaultWriteOpts); err != nil {
+		return fmt.Errorf("failed to commit PebbleDB batch: %w", err)
+	}
+
+	return nil
+}
+
+func (db *Database) Delete(storeKey string, version uint64, key []byte) error {
+	var versionBz [VersionSize]byte
+	binary.LittleEndian.PutUint64(versionBz[:], version)
+
+	batch := db.storage.NewBatch()
+
+	if err := batch.Set([]byte(latestVersionKey), versionBz[:], nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+	if err := batch.Delete(MVCCEncode(prependStoreKey(storeKey, key), version), nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+
+	if err := batch.Commit(defaultWriteOpts); err != nil {
+		return fmt.Errorf("failed to commit PebbleDB batch: %w", err)
+	}
+
+	return nil
+}
+
+func (db *Database) NewBatch(version uint64) (store.Batch, error) {
+	return NewBatch(db.storage, version)
+}
+
+// TODO: MVCC
+func (db *Database) NewIterator(storeKey string, version uint64, start, end []byte) (store.Iterator, error) {
```

Comment:
@erikgrinaker curious if you could spare some time helping me out here. I'm looking to use PebbleDB to store key/value pairs directly under a versioned scheme. I seem to have gotten CRUD to work correctly. But I'm a bit stumped on iteration.

Given a prefix and a domain, i'd like to iterate over that domain for a specific version, also handling the case where if a key wasn't updated at the version, it would simply get it from the last version it was updated at.

Reply:
So I can actually get this to work, but it iterates over _all_ versions of keys in the domain, which isn't very performant and leads me to think I'm missing something...

Reply:
Hey Bez!

Pebble doesn't really know about MVCC, it's just a plain unversioned key/value store where MVCC has to be implemented on top. As such, you'll have to iterate over all versions and implement MVCC semantics yourself.

That said, over time it has gained a number of features for working with MVCC-like data, for example the prefix/suffix concept (which corresponds to key/timestamp in MVCC terms). In particular, you may want to use [`Iterator.NextPrefix`](https://pkg.go.dev/github.com/cockroachdb/pebble#Iterator.NextPrefix) to skip to the next key with a different prefix (i.e. the latest version of the next key in MVCC terms). You may also want to try out experimental [value blocks](https://github.com/cockroachdb/pebble/blob/master/sstable/value_block.go#L21) via [`Options.Experimental.EnableValueBlocks`](https://github.com/cockroachdb/pebble/blob/58a05cd4082c6291a9c7920bcb98f1314ace98f3/options.go#L635) where historical MVCC versions are stored in separate SST blocks, to speed up read access for the latest versions.

Reply:
Super helpful, thanks @erikgrinaker! I was able to have much luck getting iteration to work with `NextPrefix`. I took a look at how cockroach does things and I found `pkg/kv/kvnemesis/engine.go`. I wonder if we can simply borrow or use this in some capacity as it has the basics of what we need. WDYT?

Reply:
O wait, I implemented `ImmediateSuccessor` and I've made progress!

Reply:
I don't believe there's a similar method for reverse iteration. But regardless, with MVCC you're typically reading at some timestamp which isn't guaranteed to be the latest timestamp, so you'll always have to do some amount of forward/reverse iteration to position correctly for each key.

Reply:
Thanks so much @erikgrinaker